src/Text.mo

```diff
@@ -26,30 +45,521 @@ module {
+    let cs = t.chars();
+    var n : Int = i;
+    loop {
+      while (n > 0) {
+        switch (cs.next()) {
+          case null assert false;
+          case (?c) n -= 1;
+        };
+      };
+      switch (cs.next()) {
+        case null assert false;
+        case (?c) { return c };
+      };
+    };
+  };
+
+  /// Returns:
+  /// - when `jo` is `null`, the subtext of `t` between characters `i` and characters `t.size()-1`.
+  ///   Traps when `t.size() < i`.
+  /// - when `jo` is  `?j`: the subtext of `t` between characters `i` and `i+j-1`.
+  ///   Traps when `t.size < i + j`.
+  public func extract(t : Text, i : Nat, jo : ?Nat) : Text {
+    var r = "";
+    let size = t.size();
+    var j = switch jo { case (?j) j; case null size };
+    if (i == 0 and j == size) return t;
+    let cs = t.chars();
+    var n = i;
+    while (n > 0) {
+      switch (cs.next()) {
+        case null (assert false);
+        case (?_) ();
+      };
+      n -= 1;
+    };
+    n := j;
+    while (n > 0) {
+      switch (cs.next()) {
+        case null (assert false);
+        case (?c) { r #= Prim.charToText(c) }
+      };
+      n -= 1;
+    };
+    return r;
+  };
+
+  /// Returns the subtext of `t` between characters `i` and `i+j-1`. Equivalent to `extract(t, i, ?j)`. Traps when `t.size < i + j`. _O_(`size(t)`).
+  public func subtext(t : Text, i : Nat, j: Nat) : Text {
+    extract(t, i, ?j);
+  };
+
+  /// Returns the concatenation of text values in `ts`.
+  public func join(ts : Iter.Iter<Text>) : Text {
+     var r = "";
+     for (t in ts) {
+       r #= t
+     };
+     return r;
+  };
+
+  /// Returns the concatenation of text values in `ts`, separated by `sep`.
+  public func joinWith(sep : Text, ts : Iter.Iter<Text>) : Text {
+    var r = "";
+    let next = ts.next;
+    switch (next()) {
+      case null { return r; };
+      case (?t) {
+        r #= t;
+      }
+    };
+    loop {
+      switch (next()) {
+        case null { return r; };
+        case (?t) {
+          r #= sep;
+          r #= t;
+        }
+      }
+    }
+  };
+
+
+  /// Returns the result of applying `f` to each character in `ts`, concatenating the intermediate single-character text values.
+  public func map(t : Text, f : Char -> Char) : Text {
+    var r = "";
+    for (c in t.chars()) {
+      r #= Prim.charToText(f(c));
+    };
+    return r;
+  };
+
+  /// Returns the result of applying `f` to each character in `ts`, concatenating the intermediate text values.
+  public func translate(t : Text, f : Char -> Text) : Text {
+    var r = "";
+    for (c in t.chars()) {
+      r #= f(c);
+    };
+    return r;
+  };
+
+
+  /// A pattern `p` describes a sequence of characters. A pattern has one of the following forms:
+  ///
+  /// * `#char c` matches the single character sequence, `c`.
+  /// * `#predicate p` matches any single character sequence `c` satisfying predicate `p(c)`.
+  /// * `#text t` matches multi-character text sequence `t`.
```

Comment:
Very nice.

Why not also have `#concat`, and perhaps even `#or`? (Future work?)

Reply:
Also, the predicate may (in some cases) want access to the "index of the character".  Just a thought for the future.

Reply:
Yeah, the itention is that we can add variants later without breaking clients. Poor man's overloading